### ISSUE TITLE
Example creating storage-optimized URIs

### ISSUE THREAD

**Issue by nyurik** [oxigraph/oxigraph#195]
Is there an example of how a user could pre-configure certain patterns for optimal URI storage?  For example, `https://www.wikidata.org/wiki/Q42` should be stored as 2 u64 (or less):  some integer representing URI prefix, and the second u64 is 42.

**Comment by Tpt** [comment]
It's not done yet, sorry. I am planning to implement it "soon" and push it into v0.4. Oxigraph encodes terms using a byte (`u8`) storing the term type and a variable amount of bytes for the value itself. I was considering using the top half of the byte (128-255) to allow at most 128 storage-optimized URI schemes.

The implementation requires a bit of refactoring of the code generating the internal RDF term representation `EncodedTerm` from the public one `Term`. It is currently implemented in a stateless function but having configuration will require injecting state into it.

The way theses patterns are configured also requires some thoughts: do we want to allow the user to provide for each URI pattern two encoding/decoding functions or do we only want to allow some pre-set encoding scheme (URI prefix + 64bit integer...)? The first method is more flexible but the second one is more configuration file friendly and would allow to easily store the configuration inside of the database to ensure the data are still readable even if for some reason the configuration file changes.

**Comment by Tpt** [comment]
>  I only saw a simple prefix:number that would need to be optimized, so I agree that doing this via configuration would make a lot of sense.

Yes, I believe also that prefix:number is the most common use cases. There are sometime other kind of small ids like [Freebase M-ids](https://developers.google.com/freebase/guide/basic_concepts#mids) but they are way less common.

I am not a huge fan of the query-based solution. It requires the SPARQL implementation to check all updates if they edit one of the meta triple. This also introduce a lot of edge cases, like should `CLEAR ALL` remove also the prefix configuration or not? Or, should this meta-triples be returned by a "SELECT "everything" operation?

I would suggest instead to provide a public Rust method named something like `add_inlined_numerical_iri_prefix` that would introduce a new prefix IRI in the database (and maybe a `remove` one). This method would be a no-op if the prefix is already known. If the prefix is new, it would store the prefix with its assigned ID into a new RocksDB table and convert the existing IRIs. 

For the stand-alone server we would have a configuration file that would provide the list of prefixes to inline and the addition method would be called on each prefix on startup. I believe we will end up to have a configuration file anyway for the server to be able e.g. to also configure if SPARQL query should target by default all graphs or only the default one. Configuration files also provide the advantage of being easily versionned into a git repository.

I believe this approach ticks all the "bonuses" of the meta-params one outside of "No need to manage separate configuration setting files".

What do you think about it?

> P.S. Do you want another issue to track this feature implementation and ideas?

We already started to discuss this on this task, maybe we could just rename it?

**Comment by nyurik** [comment]
Thanks, that makes perfect sense. I agree that setting it via config and API would make it simpler than via SPARQL as long as 128-255 byte index is not exposed, and it can be safely done multiple times and possibly even removed (last one is not really required, at least not at first).

**Comment by nyurik** [comment]
P.S. If M-IDs are required, they can always be implemented the same way as yet another API/config setting. Also, I renamed this issue.

**Comment by nyurik** [comment]
P.P.S. Updated top ticket with the summary of our discussion
